jit/core/jit_cleaner.py

```diff
@@ -0,0 +1,226 @@
+          "title", "")
+        ):
+            b["members"].remove(member)
+            binding_count += 1
+            continue
+    if not binding_count:
+        jit_logger.warn(
+          "policy binding not found. config=%s, target_project=%s, role=%s, member=%s, start=%s, end=%s",
+          conf, target_project, role, member, start, end
+        )
+        return
+    policy["bindings"][:] = [b for b in policy["bindings"] if b["members"]]
+
+    service.projects().setIamPolicy(
+      resource=target_project,
+      body={"policy": policy}
+    ).execute()
+    jit_logger.info("%d binding has been removed.", binding_count)
+    return
+
+
+def process_pubsub_msg(conf: config.JitConfig,
+  received_message: MutableSequence["ReceivedMessage"]) -> bool:
+    """
+    if message origin is not binding, we directly ack it, to clean up the pubsub
+
+    if message expired, call modify_policy_remove_member to remove the binding.
+
+    if current datetime > publishTime + 6 days
+    then republish the message, and ack original msg
+
+    if not expire, don't acknowledge,
+    as we dont expect it to appear in the queue before ack timeout.
+    :param conf:
+    :param received_message:
+    :return: is_ack bool
+    """
+
+    jit_logger.info(f"Received: {received_message.message.data}.")
+    jit_logger.info(f"attributes: {received_message.message.attributes}")
+    origin = received_message.message.attributes.get("origin", "")
+    # for invalid origin, we should directly return and make the run_jit_cleaner
+    # ack the message.
+    if not origin or constant.MessageOrigin.from_str(
+      origin) != constant.MessageOrigin.BINDING:
+        jit_logger.error(
+          f"origin invalid: origin={origin}, msg data = {received_message.message.data}.")
+        return True
+
+    # if any of these process invalid, it may raise exception,
+    # the exception will be captured in run_jit_cleaner level
+    pubsub_msg_dict = json.loads(received_message.message.data)
+    target_project_id = pubsub_msg_dict["project_id"]
+    condition = pubsub_msg_dict["conditions"]
+    expression = condition["expression"]
+    member = "user:{user}".format(user=pubsub_msg_dict["user"])
+    role = pubsub_msg_dict["role"]
+
+    start = expression.get("start", "1900-01-01T00:00:00.00000Z")
+    end = expression.get("end", "1900-01-01T00:00:00.00000Z")
+    end_datetime = datetime.strptime(end, "%Y-%m-%dT%H:%M:%S.%fZ").replace(
+      tzinfo=timezone.utc)
+
+    publish_time = received_message.message.publish_time.astimezone(
+      timezone.utc)
+    time_for_now = datetime.now(timezone.utc)
+    if end_datetime < time_for_now:
+        jit_logger.info(f"access expired: {received_message.message}")
+        modify_policy_remove_member(conf, target_project_id, role, member,
+                                    start, end)
+    elif publish_time + timedelta(days=6) < time_for_now:
+        republish(conf.pubsub_topic_name, received_message.message.data,
```

Comment:
If we republish don't we need to ACK the existing message to remove it from the queue?

Reply:
ah got it in the final return TRUE